cpp/src/communicator/mpi.cpp

```diff
@@ -183,30 +185,43 @@ std::pair<std::unique_ptr<std::vector<uint8_t>>, Rank> MPI::recv_any(Tag tag) {
     return {std::move(msg), probe_status.MPI_SOURCE};
 }
 
-std::vector<std::size_t> MPI::test_some(
-    std::vector<std::unique_ptr<Communicator::Future>> const& future_vector
+std::vector<std::unique_ptr<Communicator::Future>> MPI::test_some(
+    std::vector<std::unique_ptr<Communicator::Future>>& future_vector
 ) {
     std::vector<MPI_Request> reqs;
+    reqs.reserve(future_vector.size());
     for (auto const& future : future_vector) {
         auto mpi_future = dynamic_cast<Future const*>(future.get());
         RAPIDSMPF_EXPECTS(mpi_future != nullptr, "future isn't a MPI::Future");
         reqs.push_back(mpi_future->req_);
     }
 
     // Get completed requests as indices into `future_vector` (and `reqs`).
-    std::vector<int> completed(reqs.size());
-    {
-        int num_completed{0};
-        RAPIDSMPF_MPI(MPI_Testsome(
-            reqs.size(),
-            reqs.data(),
-            &num_completed,
-            completed.data(),
-            MPI_STATUSES_IGNORE
-        ));
-        completed.resize(static_cast<std::size_t>(num_completed));
+    std::vector<int> indices(reqs.size());
+    int num_completed{0};
+    RAPIDSMPF_MPI(MPI_Testsome(
+        reqs.size(), reqs.data(), &num_completed, indices.data(), MPI_STATUSES_IGNORE
+    ));
+    RAPIDSMPF_EXPECTS(
+        num_completed != MPI_UNDEFINED, "Expected at least one active handle."
+    );
+    if (num_completed == 0) {
+        return {};
     }
-    return std::vector<std::size_t>(completed.begin(), completed.end());
+    std::vector<std::unique_ptr<Communicator::Future>> completed;
+    completed.reserve(static_cast<std::size_t>(num_completed));
+    std::ranges::transform(
+        indices.begin(),
+        indices.begin() + num_completed,
+        std::back_inserter(completed),
+        [&](std::size_t i) {
+            std::unique_ptr<Communicator::Future> fut{nullptr};
+            std::swap(fut, future_vector[i]);
```

Comment:
Nit. Cant we do this?
```suggestion
            std::unique_ptr<Communicator::Future> fut{std::move(future_vector[i])};
```